prereise/gather/hiflddata/data_trans.py

```diff
@@ -0,0 +1,433 @@
+    return KV_dict, to_cal
+
+# Give a node, find its neighbors in 5 iteration
+
+def get_neigbors(g, node, depth=1):
+    output = {}
+    layers = dict(nx.bfs_successors(g, source=node, depth_limit=depth))
+    nodes = [node]
+    for i in range(1,depth+1):
+        output[i] = []
+        for x in nodes:
+            output[i].extend(layers.get(x,[]))
+        nodes = output[i]
+    return output
+
+# Estimate KV using neighbors
+
+def Cal_KV(N_dict,G,KV_dict,to_cal):
+    for sub in to_cal:
+        if (sub not in N_dict):
+            continue
+        neigh = get_neigbors(G, sub, depth=5)
+        temp_KV = 0
+        num = 0
+        for i in range(1, 6):
+            for nei in neigh[i]:
+                if (nei in KV_dict):
+                    temp_KV = temp_KV + KV_dict[nei]
+                    num = num + 1
+            if (num > 0):
+                KV_dict[sub] = temp_KV / num
+                continue
+            else:
+                KV_dict[sub] = -999999
+    return KV_dict
+
+
+def Set_Sub(E_csv):
+    sub_by_coord_dict = {}
+    sub_name_dict = {"sub": []}
+    raw_subs = pd.read_csv(E_csv)
+    for index, row in raw_subs.iterrows():
+        location = (row['LATITUDE'], row['LONGITUDE'])
+        if location in sub_by_coord_dict:
+            print("WARNING: substations coordinates conflict check:", location)
+        sub_by_coord_dict[location] = (row['ID'], row['NAME'])
+        if row['NAME'] not in sub_name_dict:
+            sub_name_dict[row['NAME']] = []
+        sub_name_dict[row['NAME']].append((row['LATITUDE'], row['LONGITUDE']))
+    return sub_by_coord_dict, sub_name_dict
+
+# Write sub.csv
+def Write_sub(clean_data,zone_dic):
+    sub = open('output/sub.csv','w',newline='')
+    csv_writer = csv.writer(sub)
+    csv_writer.writerow(["sub_id","sub_name","lat","lon","zone_id","type"])
+    for index, row in clean_data.iterrows():
+        csv_writer.writerow([row['ID'],row['NAME'],row['LATITUDE'],row['LONGITUDE'],zone_dic[row['STATE']],row['TYPE']])
+    sub.close()
+
+
+#Write bus.csv
+def Write_Bus(clean_data, zone_dic, KV_dict):
+    bus = open('output/bus.csv', 'w', newline='')
+    csv_writer = csv.writer(bus)
+    csv_writer.writerow(["Bus_id", "PD", "Zone_id", "base_KV"])
+    missingSub = []
+    for index, row in clean_data.iterrows():
+        sub = (row['LATITUDE'], row['LONGITUDE'])
+        if (sub in KV_dict):
+            csv_writer.writerow([row['ID'], 0, zone_dic[row['STATE']], KV_dict[sub]])
+        else:
+            missingSub.append(sub)
+    bus.close()
+    print("INFO: ", len(missingSub), " substations excluded from the network. Some examples:")
+    print(missingSub[:20])
+
+
+#Write bus2sub.csv
+def Write_bus2sub(clean_data):
+    bus2sub = open('output/bus2sub.csv','w',newline='')
+    csv_writer = csv.writer(bus2sub)
+    csv_writer.writerow(["Bus_id","sub_id"])
+    for index, row in clean_data.iterrows():
+        csv_writer.writerow([row['ID'],row['ID']])
+    bus2sub.close()
```

Comment:
If `clean_data` is a pandas DataFrame, then why do we use the low-level `csv` module and `iterrows` rather than the `DataFrame.to_csv` method?

Reply:
We were not sure the mapping between bus and substation, i.e. whether we need additional mapping info between the bus and substation so it was doing the low-level loop. In the first version, bus-to-substaion is 1: 1 mapping so only the clean_data is passed in as param. Probably prefer to leave it for now.

Reply:
will use DataFrame.to_csv in next revision.